packages/contracts-bedrock/src/cannon/libraries/MIPSMemory.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+library MIPSMemory {
+    /// @notice Reads a 32-bit value from memory.
+    /// @param _addr The address to read from.
+    /// @param _proofIndex The index of the proof in the calldata.
```

Comment:
For this function and `writeMem`, let's document here the expected proof layout in calldata. Since it's no longer obvious as these routines are in a separate library.
See https://github.com/ethereum-optimism/optimism/pull/10978/files#diff-2543eb93ac56f9072b671119906a62fec2a87591132b4315ef4745b84c8c43cbR110

Reply:
Thanks for raising this - I realized that there were a few places where these helper functions were making assumptions about the state / memory layout etc.  Reworked these helpers to try to make them agnostic to the state format.  Have a look at the last couple commits and let me know if these updates look good.

Reply:
That's better. It should be more amenable for MT-Cannon later on as we extend the witness representation.
I left a comment on additional calldata checks.